### ISSUE TITLE
Avoid typing and extending classes for transformers

### ISSUE THREAD

**Issue by ametaireau** [Kinto/kinto.js#155]
Transformers currently need to extend from a known base class. Rather than doing it this way, we could  use duck typing rather than typing, and accept any object with the correct methods.

**Comment by michielbdejong** [comment]
Yes, I was thinking the same thing. That would actually also make the whole API after #149 simpler, because we can get rid of `Kinto#createRemoteTransformer` (it might still be useful to people if we keep exposing `Kinto.transformers.RemoteTransformer`, but at least it makes it easier to define a transformer). The same would apply to `IdSchema` and `Kinto#createIdSchema` from #143.

**Comment by n1k0** [comment]
The useful thing classes add is constructors, so you can pass arguments to the created instances. A good example how it's useful is the [`TitleCharTransformer`](https://github.com/Kinto/kinto.js/blob/741b24e9a3849abdb0d5d67c8229af4b2196af7b/test/integration_test.js#L713-L724).

Sure we could keep providing a `constructor` hook to the passed object literal, though that would be quite dirty and out of the common js practices.

Of course it's still possible to pass a built object referencing some of its property, though that would harm reusability; how would you guys preserve the ability to construct transformers, and/or think of a way of achieving the very same functionality?

**Comment by michielbdejong** [comment]
The way I did it in the end for RemoteTransformers is that `kinto.createRemoteTransformer` still handles the constructor as desired, so you the TitleCharTransformer snippets from http://kintojs.readthedocs.org/en/latest/api/#transformers (both the ES6 one and the ES5 one) still work the same way, but if you want something simpler without constructor, the following is now also allowed:
````js
kinto.collection('articles', { remoteTransformers: [ {
  encode(record) { ... },
  decode(record) { ... }
} ]);
````

**Comment by n1k0** [comment]
Yeah I get that, though declaring a `constructor` prototype property is a little weird in the sense it's usually perceived as a reserved keyword, and people might expect a given behavior… Maybe that's just a minor issue. I suspect users will simply create a function to generate their own *constructed* transformers:

```js
function createTitleCharTransformer(char) {
  return {
    type: "remote",
    encode: function(record) {
      // we can use `char` here
    }
  };
}

const coll = kinto.collection("foo", {
  remoteTransformers: [ createTitleCharTransformer("?") ]
});
```

So maybe we could (should?) remove the use of ES6 classes and `constructor` entirely, to favor passing simple object literals and document the snippet above after all… That would simplify a lot the current impl. btw. Thoughts?

**Comment by ametaireau** [comment]
I agree with this proposal. Passing an object which implements the correct methods (encode and decode) seems enough for me.

**Comment by n1k0** [comment]
We'll just need to ensure provided `type` is supported, and both `encode` and `decode` are functions, and that's about it. Much simpler yet efficient, I love it.

**Comment by ametaireau** [comment]
There is something I still don't understand: why do we need to specify the `type` if we pass the object in the `remoteTransformers` option?

**Comment by n1k0** [comment]
Oh, that's true. We can definitely get rid of `type` as well :)

**Comment by ametaireau** [comment]
less is more ! Neat.

**Comment by n1k0** [comment]
As this will heavily break the API, I'm adding this to the 1.0 milestone, so we won't need to issue a new major release just after the previous has shipped.

**Comment by michielbdejong** [comment]
Great! Since the snippet from https://github.com/Kinto/kinto.js/issues/155#issuecomment-140104737 already works in master (even minus the `type: "remote"`), I'll also use that style for my blogpost about RemoteTransformers/WebCrypto (https://github.com/mozilla-services/servicedenuages.fr/pull/61).

**Comment by n1k0** [comment]
Closed by #171.
